src/app/cluster-building-blocks/QuieterReporting.h

```diff
@@ -0,0 +1,241 @@
+     */
+    AttributeDirtyState SetValue(const chip::app::DataModel::Nullable<T> & newValue, Timestamp now,
+                                 SufficientChangePredicate changedPredicate)
+    {
+        bool isChangeOfNull       = newValue.IsNull() ^ mValue.IsNull();
+        bool areBothValuesNonNull = !newValue.IsNull() && !mValue.IsNull();
+
+        bool changeToFromZero = areBothValuesNonNull && (*newValue == 0 || *mValue == 0);
+        bool isIncrement      = areBothValuesNonNull && (*newValue > *mValue);
+        bool isDecrement      = areBothValuesNonNull && (*newValue < *mValue);
+
+        bool isNewlyDirty = isChangeOfNull;
+        isNewlyDirty =
+            isNewlyDirty || (mPolicyFlags.Has(QuieterReportingPolicyEnum::kMarkDirtyOnChangeToFromZero) && changeToFromZero);
+        isNewlyDirty = isNewlyDirty || (mPolicyFlags.Has(QuieterReportingPolicyEnum::kMarkDirtyOnDecrement) && isDecrement);
+        isNewlyDirty = isNewlyDirty || (mPolicyFlags.Has(QuieterReportingPolicyEnum::kMarkDirtyOnIncrement) && isIncrement);
+
+        SufficientChangePredicateCandidate candidate{
+            mLastDirtyTimestampMillis, // lastDirtyTimestamp
+            now,                       // nowTimestamp
+            mLastDirtyValue,           // lastDirtyValue
+            newValue                   // newValue
+        };
+        isNewlyDirty = isNewlyDirty || changedPredicate(candidate);
+
+        mValue = newValue;
+
+        if (isNewlyDirty)
+        {
+            mLastDirtyValue           = newValue;
+            mLastDirtyTimestampMillis = now;
+        }
+
+        return isNewlyDirty ? AttributeDirtyState::kMustReport : AttributeDirtyState::kNoReportNeeded;
+    }
+
+    /**
+     * Same as the other `SetValue`, but assumes a changedPredicate that never overrides to dirty.
+     *
+     * This is the easy/common case.
+     *
+     * @param newValue - new value to set for the attribute
+     * @param now - system monotonic timestamp at the time of the call
+     * @return AttributeDirtyState::kMustReport if attribute must be marked dirty right away, or
+     * AttributeDirtyState::kNoReportNeeded otherwise.
+     */
+    AttributeDirtyState SetValue(const chip::app::DataModel::Nullable<T> & newValue, Timestamp now)
+    {
+        return SetValue(newValue, now, [](const SufficientChangePredicateCandidate &) -> bool { return false; });
+    }
+
+protected:
+    // Current value of the attribute.
+    chip::app::DataModel::Nullable<T> mValue;
+    // Last value that was marked as dirty (to use in comparisons for change, e.g. by SufficientChangePredicate).
+    chip::app::DataModel::Nullable<T> mLastDirtyValue;
+    // Enabled internal change detection policies.
+    QuieterReportingPolicyFlags mPolicyFlags{ 0 };
+    // Timestamp associated with the last time the attribute was marked dirty (to use in comparisons for change).
+    chip::System::Clock::Milliseconds64 mLastDirtyTimestampMillis{};
```

Comment:
We seem to be inconsistently using Milliseconds64 some places for timestamps, and Timestamp in others.  Why not just Timestamp everywhere?

Reply:
Fixed.